python/ql/src/semmle/python/objects/ObjectInternal.qll

```diff
@@ -0,0 +1,490 @@
+
+    override string toString() {
+        result = "Undefined variable"
+    }
+
+    override boolean booleanValue() {
+        none()
+    }
+
+    override ClassDecl getClassDeclaration() {
+        none()
+    }
+
+    override boolean isClass() { result = false }
+
+    override boolean isComparable() { result = false }
+
+    override ObjectInternal getClass() {
+        none()
+    }
+
+    override predicate introduced(ControlFlowNode node, PointsToContext context) {
+        none()
+    }
+
+    override Builtin getBuiltin() {
+        none()
+    }
+
+    override predicate callResult(PointsToContext callee, ObjectInternal obj, CfgOrigin origin) {
+        none()
+    }
+
+    override predicate callResult(ObjectInternal obj, CfgOrigin origin) {
+        // Accessing an undefined value raises a NameError, but if during import it probably
+        // means that we missed an import.
+        obj = ObjectInternal::unknown() and origin = CfgOrigin::unknown()
+    }
+
+    override ControlFlowNode getOrigin() {
+        none()
+    }
+
+    override int intValue() {
+        none()
+    }
+
+    override string strValue() {
+        none()
+    }
+
+    override predicate calleeAndOffset(Function scope, int paramOffset) {
+        none()
+    }
+
+    pragma [noinline] override predicate attribute(string name, ObjectInternal value, CfgOrigin origin) {
+        none()
+    }
+
+    pragma [noinline] override predicate attributesUnknown() { none() }
+
+    override predicate subscriptUnknown() { none() }
+
+    override boolean isDescriptor() { none() }
+
+    pragma [noinline] override predicate descriptorGetClass(ObjectInternal cls, ObjectInternal value, CfgOrigin origin) { none() }
+
+    pragma [noinline] override predicate descriptorGetInstance(ObjectInternal instance, ObjectInternal value, CfgOrigin origin) { none() }
+
+    pragma [noinline] override predicate binds(ObjectInternal instance, string name, ObjectInternal descriptor) { none() }
+
+    override int length() { none() }
+
+}
+
+module ObjectInternal {
+
+    ObjectInternal bool(boolean b) {
+        b = true and result = TTrue()
+        or
+        b = false and result = TFalse()
+    }
+
+    ObjectInternal none_() {
+        result = TNone()
+    }
+
+    ObjectInternal unknown() {
+        result = TUnknown()
+    }
+
+    ClassObjectInternal unknownClass() {
+        result = TUnknownClass()
+    }
+
+    ObjectInternal undefined() {
+        result = TUndefined()
+    }
+
+    ObjectInternal builtin(string name) {
+        result = TBuiltinClassObject(Builtin::builtin(name))
+        or
+        result = TBuiltinFunctionObject(Builtin::builtin(name))
+        or
+        result = TBuiltinOpaqueObject(Builtin::builtin(name))
+        or
+        name = "type" and result = TType()
+    }
+
+    ObjectInternal sysModules() {
+       result = TBuiltinOpaqueObject(Builtin::special("sys").getMember("modules"))
+    }
+
+    ObjectInternal fromInt(int n) {
```

Comment:
There should maybe be a comment that this only works for _some_ integers `n`, and that it will silently fail for all other integers.